src/dynamic_batch_scheduler.cc

```diff
@@ -170,9 +170,28 @@ DynamicBatchScheduler::~DynamicBatchScheduler()
   }
 }
 
+Status
+DynamicBatchScheduler::Update(
+    std::unique_ptr<std::lock_guard<std::mutex>>* lock)
+{
+  if (model_instance_ != nullptr) {
+    return Status(
+        Status::Code::INTERNAL,
+        "Cannot update dynamic batch scheduler specific to a model instance");
+  }
+  if (lock != nullptr) {
+    lock->reset(new std::lock_guard<std::mutex>(update_mu_));
+  }
+  return Status::Success;
+}
+
 Status
 DynamicBatchScheduler::Enqueue(std::unique_ptr<InferenceRequest>& request)
 {
+  // Pause if the scheduler is updating.
+  {
+    std::lock_guard<std::mutex> update_lock(update_mu_);
+  }
   if (stop_) {
```

Comment:
If I understand correctly - the lock only gurantees that an enqueue won't start while an update is in progress - but doesn't gurantee that an update won't happen during an enqueue - is that expected?

Reply:
Yes, it is expected, because the request(s) will be blocked by the rate limiter once the update progresses into removing the model instance, and the request can be consumed by any other instance(s) on the rate limiter after the update completes. The user [cannot specify an instance group of size zero](https://github.com/triton-inference-server/core/blob/main/src/model_config_utils.cc#L1537), so there will always be one or more instances after the update. The lock is only to block requests from rushing in during an update.

Reply:
Second thought, holding the lock on `Enqueue()` is not needed for dynamic batch scheduler, since the rate limiter is capable of handling everything, so [Remove lock on dynamic scheduler enqueue](https://github.com/triton-inference-server/core/pull/207/commits/0d426026941578a1f44d6bf09da9e5fa4957a786).

I also need to [Add gap for file timestamp update](https://github.com/triton-inference-server/server/pull/5831/commits/11db1d24af8199b78ea6a0b65c83d29b4015537a), because the code likely ran a bit faster that the 'config.pbtxt' timestamp did not get a chance to update after its content is changed, which the server thinks the model is not updated and resulted in that particular test case to fail.